api/server/app.cc

```diff
@@ -440,34 +442,58 @@ static const char *SrcCgroup() {
     return NULL;
 }
 
+void catFiles(std::string file1, std::string file2, std::string logMsg){        
+	std::ifstream ifs(file1);
+	std::ofstream ofs(file2);
+    if((ifs.is_open()) && (ofs.is_open())){
+        ofs << ifs.rdbuf();
+        ifs.close();
+        ofs.close();
+    } else {
+        printf("%s",(char*)logMsg.c_str());
+    }
+}
+
 static int InitCgroup(int multiplier) {
     const char *cgroupPath = SrcCgroup();
-
+    int check = 0;
+    
     if (!cgroupPath)
         return -1;
-    std::string create_dir("mkdir " + std::string(cgroupPath) + "/butterfly");
-
-    BASH(create_dir) {
-        LOG_WARNING_("can't create butterfly cgroup, fail cmd '%s'",
-                     create_dir.c_str());
+    std::string directory = std::string(cgroupPath) + "/butterfly" ; 
+    check = mkdir((char*)directory.c_str(),S_IWGRP );
+    if (check != 0){
+	    printf("can't create directory already exists\n");
     }
-    BASH("echo $(( `cat " + std::string(cgroupPath) + "/cpu.shares` * " +
-         std::to_string(multiplier) + " )) > " +
-         cgroupPath + "/butterfly/cpu.shares") {
-        LOG_WARNING_("can't set cgroup priority");
+    
+    std::string fileToRead = std::string(cgroupPath) + "/cpu.shares";
+    std::string fileToWrite = std::string(cgroupPath) + "/butterfly/cpu.shares";
+    std::ifstream in(fileToRead);
+    std::ofstream out(fileToWrite);
+    if ((in.is_open()) && (out.is_open())){
+	    std::string s;
+	    while (getline(in, s)){
+		    s = std::to_string(std::atoi(s.c_str()) * multiplier);
+		    out << s << std::endl;
+	    }
+	    in.close();
+	    out.close();
+    } else {
+	    LOG_WARNING_("can't set cgroup priority");
     }
-
-    // new cgroup use a diferent directory for cpushare and cpu
-    // if not, we need to initilize some files
-    BASH(std::string("! cat ") + cgroupPath + "/cpuset.mems >> /dev/null") {
-            BASH("echo `cat " + std::string(cgroupPath) + "/cpuset.mems` > " +
-                 cgroupPath + "/butterfly/cpuset.mems") {
-                    LOG_WARNING_("can't set cgroup cpuset.mems");
-            }
-            BASH("echo `cat " + std::string(cgroupPath) + "/cpuset.cpus` > " +
-                 cgroupPath + "/butterfly/cpuset.cpus") {
-                    LOG_WARNING_("can't set cgroup cpuset.cpus");
-            }
+    
+    fileToRead = std::string(cgroupPath) + "/cpuset.mems";
+    std::string msgLog;
+    if (access(fileToRead.c_str(), R_OK)){
+	    printf("access to file OK  \n");
+	    fileToWrite = std::string(cgroupPath) + "/butterfly/cpuset.mems";
+	    msgLog = "can't set cgroup cpuset.mems";
+	    catFiles(fileToRead, fileToWrite, msgLog);
+
+	    fileToRead = std::string(cgroupPath) + "/cpuset.cpus";
+	    fileToWrite = std::string(cgroupPath) + "/butterfly/cpuset.mems";
+	    msgLog = "can't set cgroup cpuset.cpus";
+	    catFiles(fileToRead, fileToWrite, msgLog);    
     }
```

Comment:
What if the access is not granted? It seems strange that there is no error logs and at the end we return 0 like if everything was ok even if not... (I apologize if I missunderstood something :) )

Reply:
> Most of the things seems ok :) Except that you deleted two comments and added no one 😢 I also added a comment about error management... Waiting for your answer!

Humm I'm not a big fan of comment in general, I see comment as an acceptation of the inability of the developer to make code self-documented(which sometime happen).

But the BASH define was super odd, here with c++ code we can have self-documented code

Reply:
though this one was useful :) so yeah maybe you should let it :)